internal/controllers/machinedeployment/machinedeployment_rollout_rollingupdate.go

```diff
@@ -292,6 +357,152 @@ func (p *rolloutPlanner) scaleDownOldMSs(ctx context.Context, totalScaleDownCoun
+// rollout is going to happen by delete/re-create, and thus it will impact availability.
+//
+// Also in place updates are assumed to impact availability, even if the in place update technically is not impacting workloads,
+// the system must account for scenarios when the operation fails, leading to remediation of the machine/unavailability.
+//
+// As a consequence:
+//   - this function can rely on scale intent previously computed, and just influence how rollout is performed.
+//   - unless the user accounts for this unavailability by setting MaxUnavailable >= 1,
+//     rollout with in-place will create one additional machine to ensure MaxUnavailable == 0 is respected.
+//
+// NOTE: if an in-place upgrade is possible and maxSurge is >= 1, creation of additional machines due to maxSurge is capped to 1 or entirely dropped.
+// Instead, creation of new machines due to scale up goes through as usual.
+func (p *rolloutPlanner) reconcileInPlaceUpdateIntent(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	if !feature.Gates.Enabled(feature.InPlaceUpdates) {
+		return nil
+	}
+
+	allMSs := append(p.oldMSs, p.newMS)
+
+	// If new MS already has all desired replicas, it does not make sense to perform more in-place updates.
+	if ptr.Deref(p.newMS.Spec.Replicas, 0) >= ptr.Deref(p.md.Spec.Replicas, 0) {
+		return nil
+	}
+
+	// Find if there are oldMSs for which it possible to perform an in-place update.
+	inPlaceUpdateCandidates := sets.Set[string]{}
+	for _, oldMS := range p.oldMSs {
+		// If the oldMS doesn't have replicas anymore, nothing left to do.
+		if ptr.Deref(oldMS.Status.Replicas, 0) <= 0 {
+			continue
+		}
+
+		// If the oldMS is not eligible for in place updates, move to the next MachineSet.
+		if result, ok := p.upToDateResults[oldMS.Name]; !ok || !result.EligibleForInPlaceUpdate {
+			continue
+		}
+
+		// Check if the MachineSet can update in place; if not, move to the next MachineSet.
+		// TODO(in-place): replace with a call to the hook
+		canUpdateMachineSetInPlaceFunc := func(_ *clusterv1.MachineSet) bool { return false }
+		if p.overrideCanUpdateMachineSetInPlace != nil {
+			canUpdateMachineSetInPlaceFunc = p.overrideCanUpdateMachineSetInPlace
+		}
+
+		canUpdateInPlace := canUpdateMachineSetInPlaceFunc(oldMS)
+		log.V(5).Info(fmt.Sprintf("CanUpdate in-place decision for MachineSet %s: %t", oldMS.Name, canUpdateInPlace), "MachineSet", klog.KObj(oldMS))
+
+		if !canUpdateInPlace {
+			continue
+		}
+
+		// Set the annotation informing the oldMS that it must move machines to the newMS instead of deleting them.
+		// Note: After a machine is moved from oldMS to newMS, the newMS will take care of the in-place upgrade process.
+		// Note: Cleanup of the MachineSetMoveMachinesToAnnotationName annotation will happen automatically as soon as the rollout planner stops
+		// to set it, because this annotation is not part of the output of computeDesiredMS
```

Comment:
If we don't have it already. Let's make sure we log whenever we change annotations on MachineSets (similar to replica numbers).

I think this will be crucial to debug race conditions (especially in situations where oldMS/newMS are changing during a rollout).

We should also make sure in future PRs to the MachineSet controller that we know through logs based on which information (probably just annotations) the MS controller made decisions to create/delete/move Machines